compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -2991,6 +2996,33 @@ object TypeComparer {
   end ApproxState
   type ApproxState = ApproxState.Repr
 
+  /** Result of `isCovered` check. */
+  object CoveredStatus:
+    opaque type Repr = Int
+
+    private inline val IsCovered = 2
+    private inline val NotHasOr = 1
+
+    /** The type is not covered. */
+    val Uncovered: Repr = 1
+
+    /** The type is covered and contains OrTypes. */
+    val CoveredWithOr: Repr = 2
+
+    /** The type is covered and free from OrTypes. */
+    val Covered: Repr = 3
+
+    object Repr:
+      extension (s: Repr)
+        def combinedWith(that: => Repr): Repr =
+          if s == Uncovered then Uncovered
+          else s min that
+
+    inline def bothHaveOr(s1: Repr, s2: Repr): Boolean = ~((s1 | s2) & NotHasOr) != 0
```

Comment:
```suggestion
    inline def bothHaveOr(s1: Repr, s2: Repr): Boolean = s1 == CoveredWithOr && s2 == CoveredWithOr 
```
Or just inline the rhs where it is used.